### PR TITLE
feat: add `--untagged-only` for independent manifest cleanup

### LINKED ISSUE #64: Purge behavior differs from documentation
**Describe the bug**
When I attempted to purge untagged images using the

```
acr purge
    --registry <Registry Name>
    --filter <Repository Name>:<Regex filter>
    --untagged
```
I received the following error `Error: required flag(s) "ago" not set`. So I added `--ago 0d` and the command went through, but it also deleted tagged images. My filter was set to `repname:.*` as I assumed the tag filter was irrelevant when you specify the `--untagged` switch. Setting the filter to `repname:^$` fixed the propblem, but the `--untagged` is still required. Omitting this switch results in no images being deleted.

So the working command was
```
acr purge
    --registry <Registry Name>
    --filter <Repository Name>:^$
    --ago 0d
    --untagged
```

### LINKED ISSUE #83: Number 999999 in the --ago key - causes the deletion of all images
**Describe the bug**
when executing the command:
acr purge --registry $REGISTRY --filter '$REPOSITORY:.*' --ago 999999d  --untagged
all images of the specified repository are deleted

**To Reproduce**
1. login in registry: acr login -r $REGISTRY.azurecr.io -u $USER -p $PASSWORD
2. execute command: acr purge --registry $REGISTRY --filter '$REPOSITORY:.*' --ago 999999d  --untagged

**Expected behavior**
when executing the command:  acr purge --registry $REGISTRY --filter '$REPOSITORY:.*' --ago 999999d  --untagged
waited for all untagged image manifests to be deleted, no time limit.
For example, if you execute the same command, but with the key:  
--ago 99999d - everything works correctly

**Screenshots**
-

**Any relevant environment information**
 - OS: Linux
 - Version: Ubuntu 20.04

**Additional context**
Suggesting to the documentation page:
https://docs.microsoft.com/ru-ru/azure/container-registry/container-registry-auto-purge
make this feature

### LINKED ISSUE #95: UX for purge is no good
The UX for this tool is no good.

It is very unclear how to use the --filter --ago and --keep to achieve what I want in a secure way.

This tool is very early still I understand that but you must have a discussion with us the users of this tool on how we use the command line switches.

Example. and what must be one of the most common use case for purge: I want to clean up all image that are not references by a tag. How do I do that? A simple "garbage collect" operation. Very unclear!

### LINKED ISSUE #480: Acr Purge --unttaged is targetting tagged manifests?
**Describe the bug**
When I run `acr purge --filter <target> --ago 1d --untagged --dry-run --registry <myRegistry>` it's saying it will delete manifests which are tagged?

**To Reproduce**
1. Have a container repository with at least 1x Tag and Multiple Manifests underneath.
2. In PowerShell, add these two variables: `$registry = "YourRegHere"` and `$repo = "YourRepoHere"`
3. Run `az acr repository show-manifests --name $registry --repository $repo --output json | ConvertFrom-Json | Select-Object digest, tags`
4. Review the output - In my case there is one tagged manifest and 30+ without tags.
5. Run `acr purge --filter $repo --ago 1d --untagged -dry-run --registry $registry`
6. Review the output - Note how the hash is in the dry run to be deleted!

**Expected behavior**
The name says --untagged so I expect it to only remove untagged manifests...

**Screenshots**
List of all manifests:
Image
List of manifests being deleted (note how the tagged one is in there)
Image

**Any relevant environment information**
 - OS: Windows
 - Version: 0.16, Revision: 2913fa43ff355e66f952f5b8b3b2a682a0fc12a4

**Additional context**
Also curious to know why these commands aren't part of the usual `az` suite? Seems a bit weird that it's a completely different executable.

### LINKED ISSUE #486: Acr Purge: Improve purge command to support untagged manifest cleanup independently
## Description:

The current implementation of the purge command introduces the --untagged flag, which enables cleanup of untagged manifests when a specific filter is applied. However, this behavior combines the cleanup of both tagged and untagged manifests, making it cumbersome to perform a cleanup of only untagged manifests.

Additionally, the date filters do not apply to untagged manifests, further complicating targeted cleanup operations.

## Expected Behavior:

Allow users to clean up untagged manifests independently of tagged ones. Ensure that filters (such as date filters) can be applied or are more clearly documented as inapplicable to untagged manifests.

## Suggested Improvement:

Introduce a dedicated flag or mode that enables exclusive cleanup of untagged manifests without requiring tagged filters. This would simplify the process and improve usability for users targeting untagged manifest cleanup only.

### LINKED ISSUE #579: ACR Purge --untagged parameter isn't respecting --ago
**Describe the bug**
Running acr purge command is showing deletion candidates for untagged manifests created less than the value specified in --ago parameter

**To Reproduce**
Run "acr-cli purge --registry *** --dry-run     --keep 20     --ago 365d     --untagged     --filter "^.*?:.*"

**Expected behavior**
The purge command should not include (untagged) manifests, if their build timestamp is younger than the timeframe specified in the --ago parameter of here 365d as described [here](https://github.com/Azure/acr-cli/blob/main/README.md#ago-flag): "You can pair --ago with the --untagged flag to apply the same age threshold to manifest cleanup, ensuring that only manifests older than the specified cutoff are removed"

**Screenshots**
acr purge output:

Image

shows a image digest which has been created on 2025-12-03T15:52:54.88223005Z

Output of "docker image inspect ..." showing image creation date:
Image

**Any relevant environment information**
 - OS: Ubuntu
 - Version of acr cli: v1.16

**Additional context**
The Azure docs describe this behaviour [here](https://learn.microsoft.com/en-us/azure/container-registry/container-registry-auto-purge): "The --untagged filter doesn't respond to the --ago filter. For additional parameters, run acr purge --help."
But I wonder why there is this mismatch to the readme in this repo here.

PS: I prefer running the acr cli compared to the ACR retention policy feature as this gives me the oppurtunity to check every single manifest deletion candidate from a --dry-run for usage in all of our kubernetes deployments. The ACR retention policy hard deletes after the timeframe.

### PR DESCRIPTION
**Pull request by balcsida**
**Purpose of the PR**

This PR introduces a new `--untagged-only` flag to the purge command, enabling users to clean up untagged manifests independently without affecting tagged manifests. This addresses the confusion and limitations reported in various issues and provides a comprehensive solution for dangling manifest cleanup.

**Key Features Added**

1. **New `--untagged-only` flag**: Exclusively targets untagged manifests without deleting any tags first
2. **Optional `--ago` and `--keep` support**: When used with `--untagged-only`, these flags become optional and provide additional filtering capabilities:
   - `--ago`: Filter untagged manifests by age (delete only manifests older than specified duration)
   - `--keep`: Preserve the most recent N untagged manifests
3. **Enhanced documentation**: Comprehensive help text and examples clearly distinguish between `--untagged` and `--untagged-only`

**Implementation Details**

- **Flag validation**: Improved validation logic with better UX (validation happens before authentication)
- **Mutual exclusivity**: `--untagged` and `--untagged-only` are mutually exclusive to prevent confusion
- **Age filtering**: Added manifest age filtering support in `purgeDanglingManifests` function
- **Keep logic**: Added support to preserve the most recent untagged manifests when `--keep` is specified
- **Comprehensive testing**: Added extensive unit tests and integration test scripts

**Enhanced User Experience**

- **Clear flag descriptions**: Updated help text clearly explains when to use each flag
- **Organized examples**: Help examples are categorized into:
  - TAG DELETION EXAMPLES (traditional workflow)
  - DANGLING MANIFEST CLEANUP EXAMPLES (new `--untagged-only` workflow)
  - ADVANCED OPTIONS
- **Better error messages**: Improved validation with clearer error messaging

**Usage Examples**

```bash
# Basic dangling manifest cleanup
acr purge -r example --untagged-only

# Clean up dangling manifests in specific repository
acr purge -r example --filter "hello-world:.*" --untagged-only

# Clean up old dangling manifests (older than 3 days), keeping 5 most recent
acr purge -r example --untagged-only --ago 3d --keep 5

# Traditional workflow: delete tags then clean up resulting dangling manifests
acr purge -r example --filter "hello-world:\w*test\w*" --ago 5d --untagged
```

**Key Differences Between Flags**

- **`--untagged`**: Used as a cleanup step *after* tag deletion. Requires `--filter` and `--ago`. Deletes tags first, then cleans up untagged manifests left behind.
- **`--untagged-only`**: Primary tool for dangling manifest cleanup. Works independently without deleting tags. Optional `--filter`, `--ago`, and `--keep` for targeted cleanup.

**Breaking Changes**
None. The existing `--untagged` flag behavior remains unchanged for backward compatibility.

**Testing**
- All existing tests pass
- Added comprehensive unit tests for new functionality
- Added integration test scripts for real Azure registry testing
- Fixed edge cases and improved error handling

Fixes #64
Fixes #83
Fixes #95
Fixes #480
Fixes #486
Fixes #579

### CODE REVIEW

**Comment by balcsida** [comment]
@estebanreyl, as the `--untagged` flag caused so much confusion in the past, do you think it would make sense to deprecate it and have some new name for this flag to prevent any accidental manifest deletion?

**Comment by estebanreyl** [comment]
I like the change, but I am little hesitant when it comes to this type of usability change scenarios, so I messaged a couple of PMs to take a look and will see what they say and will update. I am sorry its taken so long

**Comment by escherstair** [comment]
Since [this PR](https://github.com/Azure/acr-cli/pull/517) has been already merged and it has a breaking change, I think that this one should be merged ASAP too.
Because this one is the clean way to have the untagged images deleted. Otherwise everyone would creates its own workaround, playing aorund with other flags (that exist for other purposes)

File: cmd/acr/purge.go [hunk 1]
```diff
@@ -22,33 +23,40 @@ import (
...
-  - Delete all tags that are older than 7 days and begin with hello in the example.azurecr.io registry inside the hello-world repository
-    	acr purge -r example --filter "hello-world:^hello.*" --ago 7d 
-
-  - Delete all tags that are older than 7 days, begin with hello, keeping the latest 2 in example.azurecr.io registry inside the hello-world repository
+  - Delete tags older than 7 days that begin with "hello", keeping the latest 2
     	acr purge -r example --filter "hello-world:^hello.*" --ago 7d --keep 2
 
-  - Delete all tags that contain the word test in the tag name and are older than 5 days in the example.azurecr.io registry inside the hello-world 
-    repository, after that, remove the dangling manifests in the same repository
+  - Delete tags containing "test" that are older than 5 days, then clean up any dangling manifests left behind
```
Comment (estebanreyl) [inline cmd/acr/purge.go]: Ago also applies to the dangling manifests , something clearer would be something like:

"Delete tags containing ‘test’ that are older than 5 days, then delete any dangling (untagged) manifests older than 5 days"

File: cmd/acr/purge.go [hunk 2]
```diff
@@ -91,6 +100,17 @@ func newPurgeCmd(rootParams *rootParameters) *cobra.Command {
 		Long:    newPurgeCmdLongMessage,
 		Example: purgeExampleMessage,
 		RunE: func(_ *cobra.Command, _ []string) error {
+			// Validate flag combinations before authentication
+			if !purgeParams.untaggedOnly && !purgeParams.untagged {
```
Comment (estebanreyl) [inline cmd/acr/purge.go]: Doesn't --untagged still require filter?

File: cmd/acr/purge.go [hunk 3]
```diff
@@ -103,11 +123,28 @@ func newPurgeCmd(rootParams *rootParameters) *cobra.Command {
...
+			var tagFilters map[string]string
+			if purgeParams.untaggedOnly && len(purgeParams.filters) == 0 {
+				// If untagged-only without filters, get all repositories
+				allRepoNames, err := repository.GetAllRepositoryNames(ctx, acrClient.AutorestClient, purgeParams.repoPageSize)
+				if err != nil {
+					return err
+				}
+				tagFilters = make(map[string]string)
+				for _, repoName := range allRepoNames {
+					tagFilters[repoName] = ".*" // dummy filter that won't be used
```
Comment (estebanreyl) [inline cmd/acr/purge.go]: If the tag filter won't be used it seems most logical to use a filter string that cannot be matched or adjust the structure. Maybe use ^$ or the empty string?

File: cmd/acr/purge.go [hunk 4]
```diff
@@ -124,7 +161,7 @@ func newPurgeCmd(rootParams *rootParameters) *cobra.Command {
 				fmt.Printf("Specified concurrency value too large. Set to maximum value: %d \n", maxPoolSize)
 			}
 
-			deletedTagsCount, deletedManifestsCount, err := purge(ctx, acrClient, loginURL, repoParallelism, purgeParams.ago, purgeParams.keep, purgeParams.filterTimeout, purgeParams.untagged, tagFilters, purgeParams.dryRun, purgeParams.includeLocked)
+			deletedTagsCount, deletedManifestsCount, err := purge(ctx, acrClient, loginURL, repoParallelism, purgeParams.ago, purgeParams.keep, purgeParams.filterTimeout, purgeParams.untagged || purgeParams.untaggedOnly, purgeParams.untaggedOnly, tagFilters, purgeParams.dryRun, purgeParams.includeLocked)
```
Comment (estebanreyl) [inline cmd/acr/purge.go]: Maybe combine the purgeParams.untagged || purgeParams.untaggedOnly in a variable before and call it supportUntaggedCleanup? Just to make it clear, I see they are marked as mutually exclusive so just makes it clear.

File: cmd/acr/purge.go [hunk 5]
```diff
@@ -167,26 +206,44 @@ func purge(ctx context.Context,
...
 	tagFilters map[string]string,
 	dryRun bool,
 	includeLocked bool) (deletedTagsCount int, deletedManifestsCount int, err error) {
 
 	// Parse the duration once instead of for every repository
-	agoDuration, err := parseDuration(tagDeletionSince)
-	if err != nil {
-		return 0, 0, err
+	// For untagged-only mode without --ago, use 0 duration to delete all untagged manifests
+	var agoDuration time.Duration
```
Comment (estebanreyl) [inline cmd/acr/purge.go]: This should probably be done as part of the parameter validation above

File: cmd/acr/purge.go [hunk 6]
```diff
@@ -411,6 +469,24 @@ func purgeDanglingManifests(ctx context.Context, acrClient api.AcrCLIClientInter
...
 	}
 
+	// Apply keep logic if keep parameter is provided
+	if keep > 0 && len(manifestsToDelete) > keep {
+		// Sort manifests by LastUpdateTime (newest first)
+		sort.Slice(manifestsToDelete, func(i, j int) bool {
+			if manifestsToDelete[i].LastUpdateTime == nil || manifestsToDelete[j].LastUpdateTime == nil {
+				return false
+			}
+			tiI, errI := time.Parse(time.RFC3339Nano, *manifestsToDelete[i].LastUpdateTime)
```
Comment (estebanreyl) [inline cmd/acr/purge.go]: I don't expect this to happen, but this leads to slightly mysterious/inconsistent behavior in the event that we do fail to parse the last updateTime as cases where less(i, j) == false and less(j, i) == false can happen without the items being equal (because you haven’t actually compared anything). I would suggest establishing a consistent ordering (for example the nil one is always considered older) and maybe moving the keep functionality to its own function so it can be tested independently.  I asked copilot to do it and it gave me this:

```go
sort.Slice(manifestsToDelete, func(i, j int) bool {
    mi := manifestsToDelete[i]
    mj := manifestsToDelete[j]

    // Extract strings; nil => invalid
    var si, sj string
    if mi.LastUpdateTime != nil {
        si = *mi.LastUpdateTime
    }
    if mj.LastUpdateTime != nil {
        sj = *mj.LastUpdateTime
    }

    ti, errI := time.Parse(time.RFC3339Nano, si)
    if errI != nil && si != "" {
        // fallback to RFC3339 if needed
        if t, err := time.Parse(time.RFC3339, si); err == nil {
            ti, errI = t, nil
        }
    }
    tj, errJ := time.Parse(time.RFC3339Nano, sj)
    if errJ != nil && sj != "" {
        if t, err := time.Parse(time.RFC3339, sj); err == nil {
            tj, errJ = t, nil
        }
    }

    // Define validity flags
    vi := (errI == nil)
    vj := (errJ == nil)

    // Ordering rules (total order):
    // 1) Valid timestamps come before invalid (invalid considered oldest).
    if vi != vj {
        return vi // true if i valid and j invalid
    }

    // 2) Both valid: newest first.
    if vi && vj {
        if ti.Equal(tj) {
            // 3) Tie-breaker for determinism (adjust to your field)
            return mi.Digest < mj.Digest
        }
        return ti.After(tj) // newest first
    }

    // 4) Both invalid: tie-breaker for determinism
    return mi.Digest < mj.Digest
})
```

File: scripts/experimental/test-purge-untagged-only.sh [hunk 7]
```diff
@@ -0,0 +1,653 @@
...
+    echo "Testing --untagged-only performance..."
+    local start_time=$(date +%s%N)
+    "$ACR_CLI" purge --registry "$REGISTRY" --filter "$repo:.*" --untagged-only --dry-run >/dev/null 2>&1
+    local end_time=$(date +%s%N)
+    local untagged_only_time=$((end_time - start_time))
+    
+    # Test regular purge with --untagged performance
+    echo "Testing regular purge with --untagged performance..."
+    start_time=$(date +%s%N)
+    "$ACR_CLI" purge --registry "$REGISTRY" --filter "$repo:.*" --ago 0d --untagged --dry-run >/dev/null 2>&1
```
Comment (estebanreyl) [inline scripts/experimental/test-purge-untagged-only.sh]: to have a real comparison shouldn't you avoid cleaning up any tags, so use 0^$ here?

**Review by estebanreyl — COMMENTED** [review]
Overall LGTM, we are in line with including the changes. I've added some comments on some small items that would be nice to address before merge. Thanks again for the contribution!

File: cmd/acr/purge.go (hunk 3 above)
Comment (balcsida) [reply cmd/acr/purge.go]: Changed to empty string

File: cmd/acr/purge.go [hunk 8]
```diff
@@ -143,19 +180,21 @@ func newPurgeCmd(rootParams *rootParameters) *cobra.Command {
 		},
 	}
 
-	cmd.Flags().BoolVar(&purgeParams.untagged, "untagged", false, "If the untagged flag is set all the manifests that do not have any tags associated to them will be also purged, except if they belong to a manifest list that contains at least one tag")
+	cmd.Flags().BoolVar(&purgeParams.untagged, "untagged", false, "In addition to deleting tags (based on --filter and --ago), also delete untagged manifests that were left behind after tag deletion. This is typically used as a cleanup step after deleting tags. Note: This requires --filter and --ago to be specified")
+	cmd.Flags().BoolVar(&purgeParams.untaggedOnly, "untagged-only", false, "Clean up dangling manifests: Delete ONLY untagged manifests (manifests without any tags), without deleting any tags first. This is the primary way to clean up dangling manifests in your registry. Optional: Use --ago to delete only old untagged manifests, --keep to preserve recent ones, and --filter to target specific repositories")
```
[the remark this answers is not in the document]
Comment (balcsida) [reply cmd/acr/purge.go]: Good catch, added note to flag description
